src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+	//   expected tracked cooldowns per window is variable
+	protected getBaselineExpectedTrackedCooldown(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return cooldown.expectedPerWindow || 0
+	}
+
+	// Override point for class-specific logic to change expected uses of a tracked cooldown per BuffWindow - default no effect
+	// Return a positive number to increase expected GCDs for this window, and a negative number to decrease
+	protected changeExpectedTrackedCooldownClassLogic(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return 0
+	}
+
+	private getBuffWindowExpectedGCDs(buffWindow: BuffWindowState): number {
+		return this.getBaselineExpectedGCDs(buffWindow) + this.changeExpectedGCDsClassLogic(buffWindow) - this.reduceExpectedGCDsEndOfFight(buffWindow)
+	}
+
+	private getBuffWindowExpectedTrackedCooldowns(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return this.getBaselineExpectedTrackedCooldown(buffWindow, cooldown) + this.changeExpectedTrackedCooldownClassLogic(buffWindow, cooldown)
+	}
+
+	private onComplete() {
+		if ( this.buffAction && this.expectedGCDs ) {
+			const missedGCDs = this.buffWindows
+				.reduce((sum, curWindow) => {
+					const expectedGCDs = this.getBuffWindowExpectedGCDs(curWindow)
+					return sum + Math.max(0, expectedGCDs - curWindow.gcds)
+				}, 0)
+
+			this.suggestions.add(new TieredSuggestion({
+				icon: this.buffAction.icon,
+				content: this.expectedGCDs.suggestionContent,
+				tiers: this.expectedGCDs.severityTiers,
+				value: missedGCDs,
+				why: <Trans id="core.buffwindow.suggestions.missedgcd.why">
+					{missedGCDs} <Plural value={missedGCDs} one="GCD was" other="GCDs were" /> missed during {this.buffAction.name} windows.
+				</Trans>,
+			}))
+		}
+
+		if ( this.buffAction && this.requiredGCDs ) {
+			const allowedGCDsById = this.requiredGCDs.actions.map(a => a.id)
+			const invalidGCDs = this.buffWindows
+				.reduce((sum, curWindow) => sum + Math.max(0, curWindow.gcds - curWindow.getActionCountByIds(allowedGCDsById)), 0)
+
+			this.suggestions.add(new TieredSuggestion({
+				icon: this.requiredGCDs.iconAction.icon,
+				content: this.requiredGCDs.suggestionContent,
+				tiers: this.requiredGCDs.severityTiers,
+				value: invalidGCDs,
+				why: <Trans id="core.buffwindow.suggestions.badgcd.why">
+					{invalidGCDs} incorrect <Plural value={invalidGCDs} one="GCD was" other="GCDs were" /> used during {this.buffAction.name} windows.
+				</Trans>,
+			}))
+		}
+
+		if ( this.trackedCooldowns ) {
+			this.trackedCooldowns.forEach(cooldown => {
+				if ( this.buffAction ) {
+					const missedCooldowns = this.buffWindows
+						.reduce((sum, curWindow) => sum + Math.max(0, cooldown.expectedPerWindow - curWindow.getActionCountByIds([cooldown.action.id])), 0)
+
+					this.suggestions.add(new TieredSuggestion({
```

Comment:
By doing this in a loop, there's a real chance of generating a _massive_ number of suggestions.

I'd _generally_ prefer these to be grouped as a single thing - "You missed x important skills" rather than "You missed y foo" + "You missed z bar" + ...

Reply:
I implemented this way to duplicate the way Warrior's IR is currently doing it.  Not opposed to changing, but would need some thought as to how that's going to impact the output for IR.

Reply:
Ping me when that's resolved, i remember seeing a convo about it on discord.

Reply:
Per discussion with Skye, went ahead and set trackedCooldowns / trackedBadCooldowns to collapse to a single suggestion regardless of number of cooldowns being tracked.

I also updated the naming to consistently use "Action" instead of "Cooldown", since potentially there may be implementers that want to track GCDs as well as oGCDs (e.g. Fight or Flight x Goring Blade)